import os,re,tarfile,time,subprocess
#glinet_version_update  where tmps is mounted in /tmp  when you write to /tmp  tou write in ram
#note : to make an update file  you have to make a tar of the onos files inside the  script_folder  location where webserver.py is ...
#then rename the tar to scripts_folder.tar.gz
#don't make a tar from the previous directory!
#then update the online md5sum for the script and for the tar!



tmp_dir="/tmp/onos_update/"


logprint("current_hw:"+current_hw)
logprint("base_path:"+base_path)   # /bin/onos/  or ""
logprint("current_local_fw"+current_local_fw)
logprint("online_fw_folder_url:"+online_fw_folder_url)   # get from the update_check.py that execute this script
logprint("server_url:"+online_fw_folder_url) 
logprint("router_hardware_type:"+router_hardware_type)
logprint("os.getcwd():"+os.getcwd())
logprint("server_url:"+server_url)  # get from the update_check.py that execute this script

url_php=server_url+"updates/mail_update.php"

msg="ex0"
s_mail="1"
send_to_php_log(router_sn,current_hw,current_local_fw,msg,s_mail,url_php)


free_ram_space=0
try:
  free_ram_space=float((re.search('tmpfs(.+?)\n',os.popen("df").read().split("/tmp ")[0]).group(1)).split()[2]) #get free ram space in kbytes
except :
  logprint("error , cannot read free memory with df cmd")


if (free_ram_space < 4000): #if there are not at least 4 mbyte of free space in ram ...
  logprint("error the hardware has not enaught free ram space")
  msg="errRam"
  s_mail="1"
  send_to_php_log(router_sn,current_hw,current_local_fw,msg,s_mail,url_php)
  quit()
else: # there is enough free memory
  #os.system("touch va.txt")
  #http://myonos.com/onos/updates/glinet/scripts_folder.tar.gz


  if online_fw_folder_url=="local":  #the files are already in the onos_update/ folder
    logprint("local folder used for update")
    try:
      md5_file = open("onos_update/md5sum_tar.txt",'r')  
      md5_code=(md5_file.read()).strip()   #get the string and delete the \n
      md5_file.close()

      file_md5code=os.popen("md5sum onos_update/scripts_folder.tar.gz").read().split(" ")[0] 
      if md5_code!=file_md5code:
        logprint("0 update failed,corrupted tar")
        msg="errlocal"
        s_mail="0"
        send_to_php_log(router_sn,current_hw,current_local_fw,msg,s_mail,url_php)
        os.system("rm -f onos_update/")  # delete the folder so next time it will be redownloaded by update_check.py
        quit()
    except Exception as e:     
      message="error, maybe there is not internet connection"
      logprint(message)
      logprint(e)
      logprint(sys.exc_info())

 

  else:  # I need to download  the files..
    try_number=5  
    while (try_number>0): #retry 5 times to retrieve the file before to giveup
      #tar_file= urllib2.urlopen(onos_online_server_url+'updates/'+current_hw+'/scripts_folder.tar.gz').read() 
      md5sum_tar_url=online_fw_folder_url+'''/md5sum_tar.txt '''
      scripts_folder_tar_url=online_fw_folder_url+'''/scripts_folder.tar.gz'''
      logprint("i download md5sum_tar from:"+md5sum_tar_url) 

      logprint("i download scripts_folder_tar from:"+scripts_folder_tar_url) 

      logprint("rm -rf "+tmp_dir)
      try:
        subprocess.call("rm -rf "+tmp_dir, shell=True,close_fds=True)
      except:
        pass

      try:
        subprocess.call("mkdir -p "+tmp_dir, shell=True,close_fds=True)
        subprocess.call('''wget -O '''+tmp_dir+'''md5sum_tar.txt '''+md5sum_tar_url , shell=True,close_fds=True)
        subprocess.call('''wget -O '''+tmp_dir+'''scripts_folder.tar.gz '''+scripts_folder_tar_url, shell=True,close_fds=True)


        md5_file = open(tmp_dir+"md5sum_tar.txt",'r')  
        md5_code=(md5_file.read()).strip()   #get the string and delete the \n
        md5_file.close()

        file_md5code=os.popen("md5sum "+tmp_dir+"scripts_folder.tar.gz").read().split(" ")[0] 
        if md5_code!=file_md5code:
          print "file is corrupted"
          if try_number==1:
            print "update failed,corrupted tar"
            msg="errUpMd5"
            s_mail="0"
            send_to_php_log(router_sn,current_hw,current_local_fw,msg,s_mail,url_php)
            quit()
          else:
            try_number=try_number-1
            continue
            
 
        else: #the file is not corrupted
          print "tar is ok"
          break

      #fo = open("/tmp/onos_update/scripts_folder.tar.gz", "wb")
      #fo.write(tar_file)
      #tar_file.close()
      #fo.close()  
      except Exception as e :
        logprint("some error happened in the scripts_folder.tar.gz download")  
        logprint(e.args)
        try_number=try_number-1









  #here i have the right tar in  /tmp/onos/onos_update/

  subprocess.call("mkdir -p "+tmp_dir+"downloaded_fw", shell=True,close_fds=True)

  subprocess.call("tar xzf "+tmp_dir+"/scripts_folder.tar.gz -C "+tmp_dir+"downloaded_fw/", shell=True,close_fds=True)

#rm -R /tmp/ram/b/onos_5.15/scripts_folder/!(cgi|zone|config_files|zones|globalVar.py)
  #print ("rm -R "+base_path+"scripts_folder/!(cgi|zone|config_files|zones|globalVar.py)")
  #os.system('''rm -rf '''+base_path+'''scripts_folder/!(cgi|zone|config_files|zones|globalVar.py)''')

  
  logprint ('''ls -1 '''+base_path+'''scripts_folder/ | grep -v "csv\|cgi\|config_files" | xargs rm -rf ''')
  #os.system('''cd '''+base_path+'''scripts_folder/''')
  #os.system('''cd '''+base_path+'''scripts_folder/ && ls -1 | grep -v "cgi\|config_files" | xargs rm -rf ''')

  subprocess.call('''ls -1 '''+base_path+'''scripts_folder/ | grep -v "csv\|cgi\|config_files" | xargs rm -rf ''', shell=True,close_fds=True)

  logprint ('''cp -r -n'''+tmp_dir+'''downloaded_fw/scripts_folder/* '''+base_path+'''scripts_folder/''')

  subprocess.call('''cp -r -n '''+tmp_dir+'''downloaded_fw/scripts_folder/* '''+base_path+'''scripts_folder/''', shell=True,close_fds=True)
  logprint("clear directories")

  if online_fw_folder_url!="local":
    fw_version_url=online_fw_folder_url+'''/fw_version.txt''' 
    logprint("I get fw_version.txt from:"+fw_version_url)
    subprocess.call('''wget -O '''+base_path+'''fw_version.txt '''+fw_version_url, shell=True,close_fds=True)
    time.sleep(10)
    subprocess.call('''rm -rf '''+base_path+'''onos_update''', shell=True,close_fds=True)
  else:

    os.system('''cp onos_update/fw_version.txt '''+base_path+'''fw_version.txt ''')
    os.system('''rm -rf onos_update''')



#here the update was successful....
#i will check if the folder contain an arduino .hex file to update the new serial_interface.

#avrdude -v -p atmega328p -c arduino -P /dev/ttyUSB1 -b 115200 -D -U flash:w:/tmp/build8621467179970249087.tmp/target.cpp.hex:i
#the folder is scripts_folder/arduino_update/
hex_arduino_update_path=base_path+"scripts_folder/arduino_update/onos_rfm69_serial_interface.hex"

logprint("os.getcwd():"+os.getcwd())
if os.path.isfile(hex_arduino_update_path):  #if the directory and the file exist...
  if router_hardware_type=="RouterOP": #serial port is s2 in this hw
    flash_cmd='''avrdude -v -p atmega328p -c arduino -P /dev/ttyS2 -b 115200 -D -U flash:w:'''+hex_arduino_update_path+''':i'''
# avrdude -v -p atmega328p -c arduino -P /dev/ttyS2 -b 115200 -D -U flash:w:scripts_folder/arduino_update/onos_rfm69_serial_interface.hex:i

#sudo avrdude -p atmega328p -C avrdude_gpio.conf -c pi_1 -v -U flash:w:scripts_folder/arduino_update/onos_rfm69_serial_interface.hex:i

    try:
      subprocess.call(flash_cmd, shell=True,close_fds=True)
    except Exception as e :
      logprint("some error happened in the arduino flashing")  
      logprint(e.args)

  else:
    logprint("hw not supported for arduino update")

else:
  logprint("no file to update the arduino with")






